### PR TITLE
changes related to ROI random selection in MLT

### PR DESCRIPTION
**Pull request by MRiganSUSX**
This is a proposed solution to [Trigger Issue 230](https://github.com/DUNE-DAQ/trigger/issues/230) as part of the 4.2.0 trigger deliverables for ROI selection in MLT. More details in the aforementioned issue.

### CODE REVIEW

File: python/daqconf/apps/trigger_gen.py [hunk 1]
```diff
@@ -91,7 +91,18 @@ def get_trigger_bitwords(bitwords):
...
+def check_mlt_roi_config(mlt_roi_conf, n_groups):
+    prob = 0
+    for group in mlt_roi_conf:
+        prob += group["probability"]
+        if group["number_of_link_groups"] > n_groups:
+            raise RuntimeError(f'The MLT ROI configuration map is invalid, the number of requested link groups must be <= all link groups')
+    if prob > 1.0:
+        raise RuntimeError(f'The MLT ROI configuration map is invalid, the sum of probabilites must be <= 1.0')
+    return
+
```
Comment (ArturSztuc) [inline python/daqconf/apps/trigger_gen.py:104]: I don't understand why the sum of probabilities must be <= 1? Perhaps I don't understand how this random MLT ROI is supposed to work, or what this function is supposed to do exactly.
I imagine that if you have e.g.
```
{
                "groups_selection_mode": "kRandom",
                "number_of_link_groups": 100,
                "probability": 0.8,
}
```
means that each of the 100 links has 80% of being "triggered" with some random selection algo.
You can imagine a scenario:
```
{
                "groups_selection_mode": "kRandom",
                "number_of_link_groups": 100,
                "probability": 0.8,
},
{
                "groups_selection_mode": "kFunkySelectionAlgo",
                "number_of_link_groups": 100,
                "probability": 0.5,
}
```

Which should be valid too, and will trigger P(A or B) = P(A) + P(B) - P(A and B), so 0.8 + 0.5 - (0.5*0.8) = 0.9 = 90% of links.

File: python/daqconf/core/fragment_producers.py [hunk 2]
```diff
@@ -25,16 +26,77 @@ def set_mlt_links(the_system, mlt_app_name="trigger", verbose=False):
...
+                matched = False
+                for key in mlt_readout_map.keys():
+                    if key != -1:
+                        if source_id.id in mlt_readout_map[key]["elements"]:
+                            mlt_links["groups"][key]["links"].append( mlt.SourceID(subsystem=ensure_subsystem_string(source_id.subsystem), element=source_id.id) )
+                            matched = True
+                # special case to cover readout that is not a TP source for trigger
+                if not matched:
+                    mlt_links["groups"].append({"group": len(mlt_links["groups"]), "links": []})
+                    mlt_links["groups"][len(mlt_links["groups"])-1]["links"].append( mlt.SourceID(subsystem=ensure_subsystem_string(source_id.subsystem), element=source_id.id) )
```
Comment (ArturSztuc) [inline python/daqconf/core/fragment_producers.py]: No "mandatory" case here? Is there documentation about this that I could read?

**Review by ArturSztuc — APPROVED** [review]
Looks good to me! I have comments/questions rather than request changes.
Could you link some presentation/slides/documentation to this PR?

File: python/daqconf/core/fragment_producers.py (hunk 2 above)
Comment (MRiganSUSX) [reply python/daqconf/core/fragment_producers.py]: All currently existing documentation is here: [trigger Issue 230](https://github.com/DUNE-DAQ/trigger/issues/230)

File: python/daqconf/apps/trigger_gen.py (hunk 1 above)
Comment (MRiganSUSX) [reply python/daqconf/apps/trigger_gen.py:104]: This is for simplicity, ie we have a 100% chance of reading out and we divide that between the configuration groups. You are right though that there is a different approach that could have been taken, treating these groups separately. However, then we would need to keep track of groups that have been read out (as to not read them out multiple times for the same TD). Maybe we can do that, but not for now.

**Comment by bieryAtFnal** [comment]
Regression tests mostly show success, however,  when TPG is enabled and there are several data sources, I see some errors.

Here are steps to reproduce what I see on an AL9 computer:
```
source /cvmfs/dunedaq.opensciencegrid.org/setup_dunedaq.sh
setup_dbt latest
dbt-create -c -n NAFD23-10-14 14OctFDDevMLTROITest
cd 14OctFDDevMLTROITest/sourcecode
git clone https://github.com/DUNE-DAQ/daqconf -b mrigan/mlt_roi_rnd
git clone https://github.com/DUNE-DAQ/fddaqconf -b mrigan/mlt_roi_rnd
git clone https://github.com/DUNE-DAQ/trigger -b mrigan/mlt_roi_rnd
cd ..
dbt-workarea-env
git clone https://github.com/DUNE-DAQ/nanorc -b develop
cd nanorc
pip install -U .
cd ..
dbt-build -j 12
dbt-workarea-env

mkdir rundir
cd rundir

cat <<EOF1 > daqconf.json
{
  "boot": {
    "use_connectivity_service": true,
    "start_connectivity_service": true,
    "connectivity_service_host": "localhost",
    "connectivity_service_port": 15432
  }, 
  "daq_common": {
    "data_rate_slowdown_factor": 1
  },
  "detector": {
    "clock_speed_hz": 62500000
  },
  "readout": {
    "use_fake_cards": true,
    "default_data_file": "asset://?checksum=dd156b4895f1b06a06b6ff38e37bd798",
    "generate_periodic_adc_pattern": true,
    "emulated_TP_rate_per_ch": 1,
    "enable_tpg": true,
    "tpg_threshold": 500,
    "tpg_algorithm": "SimpleThreshold"
  },
  "trigger": {
    "trigger_window_before_ticks": 1000,
    "trigger_window_after_ticks": 1000,
    "trigger_activity_config": {
      "prescale": 25
    },
    "mlt_merge_overlapping_tcs": false
  },
  "dataflow": {
    "apps": [ { "app_name": "dataflow0" } ],
    "enable_tpset_writing": true,
    "token_count": 20
  },
  "hsi": {
    "random_trigger_rate_hz": 1.0
  }
}
EOF1

cat <<EOF2 > dro_map.json
[
    {
        "src_id": 0,
        "geo_id": {
            "det_id": 3,
            "crate_id": 1,
            "slot_id": 0,
            "stream_id": 0
        },
        "kind": "eth",
        "parameters": {
            "protocol": "udp",
            "mode": "fix_rate",
            "rx_iface": 0,
            "rx_host": "localhost",
            "rx_mac": "00:00:00:00:00:00",
            "rx_ip": "0.0.0.0",
            "tx_host": "localhost",
            "tx_mac": "00:00:00:00:00:00",
            "tx_ip": "0.0.0.0"
        }
    },
    {
        "src_id": 1,
        "geo_id": {
            "det_id": 3,
            "crate_id": 1,
            "slot_id": 0,
            "stream_id": 1
        },
        "kind": "eth",
        "parameters": {
            "protocol": "udp",
            "mode": "fix_rate",
            "rx_iface": 0,
            "rx_host": "localhost",
            "rx_mac": "00:00:00:00:00:00",
            "rx_ip": "0.0.0.0",
            "tx_host": "localhost",
            "tx_mac": "00:00:00:00:00:00",
            "tx_ip": "0.0.0.0"
        }
    },
    {
        "src_id": 2,
        "geo_id": {
            "det_id": 3,
            "crate_id": 2,
            "slot_id": 0,
            "stream_id": 0
        },
        "kind": "eth",
        "parameters": {
            "protocol": "udp",
            "mode": "fix_rate",
            "rx_iface": 1,
            "rx_host": "localhost",
            "rx_mac": "00:00:00:00:00:01",
            "rx_ip": "0.0.0.1",
            "tx_host": "localhost",
            "tx_mac": "00:00:00:00:00:00",
            "tx_ip": "0.0.0.0"
        }
    },
    {
        "src_id": 3,
        "geo_id": {
            "det_id": 3,
            "crate_id": 2,
            "slot_id": 0,
            "stream_id": 1
        },
        "kind": "eth",
        "parameters": {
            "protocol": "udp",
            "mode": "fix_rate",
            "rx_iface": 1,
            "rx_host": "localhost",
            "rx_mac": "00:00:00:00:00:01",
            "rx_ip": "0.0.0.1",
            "tx_host": "localhost",
            "tx_mac": "00:00:00:00:00:00",
            "tx_ip": "0.0.0.0"
        }
    },
    {
        "src_id": 4,
        "geo_id": {
            "det_id": 3,
            "crate_id": 3,
            "slot_id": 0,
            "stream_id": 0
        },
        "kind": "eth",
        "parameters": {
            "protocol": "udp",
            "mode": "fix_rate",
            "rx_iface": 2,
            "rx_host": "localhost",
            "rx_mac": "00:00:00:00:00:02",
            "rx_ip": "0.0.0.2",
            "tx_host": "localhost",
            "tx_mac": "00:00:00:00:00:00",
            "tx_ip": "0.0.0.0"
        }
    },
    {
        "src_id": 5,
        "geo_id": {
            "det_id": 3,
            "crate_id": 3,
            "slot_id": 0,
            "stream_id": 1
        },
        "kind": "eth",
        "parameters": {
            "protocol": "udp",
            "mode": "fix_rate",
            "rx_iface": 2,
            "rx_host": "localhost",
            "rx_mac": "00:00:00:00:00:02",
            "rx_ip": "0.0.0.2",
            "tx_host": "localhost",
            "tx_mac": "00:00:00:00:00:00",
            "tx_ip": "0.0.0.0"
        }
    }
]
EOF2

fddaqconf_gen -c ./daqconf.json --detector-readout-map-file ./dro_map.json mdapp_wibeth_tpg
nanorc mdapp_wibeth_tpg ${USER}-test boot conf start_run 101 wait 20 stop_run scrap terminate

egrep 'ERROR|WARNING' log_*.txt
```

**Review by bieryAtFnal — APPROVED** [review]
LGTM

**Pull request merged**